### ISSUE TITLE
Game Plan

### ISSUE THREAD

**Issue by chacha** [Team3128/robot-course-description#1]
So, we have a course description. What are the next steps that we need to make this Robot Course a reality?

My ideas are:

1. Identify key robot-oriented concepts
2. Create lessons for teaching those concepts to programmers
3. Execute
4. Get Feedback
5. Repeat Step 2

Thoughts?

**Comment by multiplemonomials** [comment]
Right, so I guess we could divide what the class teaches into three sections: 
- Supplemental java things that people will need to know in order to work with our code (threads, unit tests, debugging, etc.)
- How to actually write programs for the robot.  Mainly, this will be about showing them how to get the SDK installed and how to deploy the code onto the RoboRIO.  I may also cover the emulator.
- How to use the robot APIs.  This includes the WPILib API as well as our robot codebase.  Probably, I will assign them to configure the code for a robot (create new MainClass subclasses) for practice and test it on the new practice bot or the emulator.

**Comment by multiplemonomials** [comment]
Going down a level of recursion, the things we need to teach in the supplemental Java course are, according to public opinion:
- Debugging (this is covered in AP CompSci, but very briefly)
- Lambdas
- Function pointers
- Threads, synchronized blocks, and multithreading
- Generics (people should know how to use them, but I want to cover how to make them)
- Interfaces (I'm not exactly sure how much of this I need to cover)
- The testing mentality (I need to be better about this myself if I don't want to be a hypocrite when teaching this)

Other stuff we may want to to cover:
- Bitwise operations
- Enums
- Semaphores

**Comment by multiplemonomials** [comment]
For the robot programming part:
- a general introduction to the control system and the robot control hardware (RoboRIO, PDP, VRM, battery, etc.)
- installing the driver station
- setting up Eclipse and the FRC plugin
- creating a hello world project and deploying code to the RoboRIO
- debugging code on the RoboRIO
- how ports on the RoboRIO work
- a discussion of the cRIO and how to work with it. (After Kian graduates, me, Raylen, and maybe Jacob will be the only ones who know how to write code for the cRIO)

Possible additions:
- Reimaging the RoboRIO
- ssh-ing to the RoboRIO
- a discussion of why the cRIO was so bad (it is very slow and has poor java support) and why the RoboRIO is so much better

**Comment by multiplemonomials** [comment]
Lessons for the API part:
1. Cloning and modifying our codebase, and using the MainClass system.
2. Overview of the hardware classes, and a test robot using the emulator.
3. Coding for an actual test robot
4. Writing new hardware classes and unit testing them.

**Comment by chacha** [comment]
Depending upon the pricing, might it just be a better idea to ditch the cRIO all together? @Garrison-Price Any thoughts on this?
